### PR TITLE
Tweet when a new repository is added

### PR DESCRIPTION
**Pull request by sam95**
To be installed :

     pip install twython
TO-DOs
1. Shorten the Repo-URL.
2. Patch in the credentials for the twitter handle.
3. Backfill for existing repos.

### CODE REVIEW

File: gfi/populate.py [hunk 1]
```diff
@@ -117,13 +122,15 @@ def get_repository_info(owner, name):
...
         LOGGER.info("Found %d repository entries in %s", len(DATA["repositories"]), REPO_DATA_FILE)
-
+        twitter = Twython(APP_KEY, APP_SECRET, OAUTH_TOKEN, OAUTH_TOKEN_SECRET)
         for repository_url in DATA["repositories"]:
             repo_dict = parse_github_url(repository_url)
             if repo_dict:
                 repo_details = get_repository_info(repo_dict["owner"], repo_dict["name"])
                 if repo_details:
                     REPOSITORIES.append(repo_details)
+                    tweet_string = TWEET_TEMPLATE.substitute(owner = repo_dict["owner"], url = repo_dict["url"])
```
Comment (jaipradeesh) [inline gfi/populate.py]: https://twitter.com/gfibot/status/1227011275260977153 -- Can we use this format?

If there is a way to read the project's description.

Screenshot 2020-02-21 at 2 09 56 AM

Comment (sam95) [reply gfi/populate.py]: Created a template and added the same with the PR. Do review and let me know if I'm missing anything.

**Comment by sam95** [comment]
Hi Rahul,

That's what I had mentioned in the issue thread. We can also look at maintaining a persistent REDIS. 
Meanwhile I will also explore if we can maintain this through twython layer.

**Comment by sam95** [comment]
Hi @RJ722,

This already seems to have been handled by twitter APIs themselves. We will not be able to make duplicate status updates.

**For each update attempt, the update text is compared with the authenticating user's recent Tweets. Any attempt that would result in duplication will be blocked, resulting in a 403 error.**

Reference : [https://developer.twitter.com/en/docs/tweets/post-and-engage/api-reference/post-statuses-update](url)

However, we will still need to monitor if this happens over a period of time (like a week).

**Comment by jaipradeesh** [comment]
Couple of things to keep in mind:
- We might add a lot of repositories in on go. This shouldn't spam our followers' twitter feed
- If Twitter already handles duplicate tweet prevention, we should randomly pick repositories every day and tweet it.

@sam95 @RJ722

**Comment by RJ722** [comment]
> For each update attempt, the update text is compared with the authenticating user's recent Tweets. Any attempt that would result in duplication will be blocked, resulting in a 403 error.

Hmm, interesting!

I'm not sure how the define 'recent' - like 100 ms recent, or an 'hour' recent? In any case, we shouldn't be relying on Twitter policies for the bot - they can change anytime, and are only meant for regulation.

I'd recommend writing a different script (and maintaining it in the same repo, here) which can use a very light weight `.sqlite` database (which we can host on DeepSource infrastructure).

**Comment by jaipradeesh** [comment]
Hi @sam95 @RJ722 any update on this?

**Comment by sam95** [comment]
Hi @jaipradeesh,

Was out of town for 2 weeks. We will have this closed and merged by the End of this week.

**Comment by sam95** [comment]
Hi @jaipradeesh and @RJ722,

Please review the code and let me know if there are changes required. I have tested this on my local system as well and it works as expected.

**Comment by sam95** [comment]
Hi @RJ722 and @jaipradeesh ,

Thanks for the review. The code looks much cleaner now. Let me know if there are more changes required.

Thanks and Regards,
Sameera

**Comment by RJ722** [comment]
Other than this, we'd also need to add new dependencies to `pyproject.toml` and regenerate `poetry.lock`.

**Comment by RJ722** [comment]
@sam95 I thought about this some more.

Now, I know that you've worked quite a lot on improving the database thing, but here's a much simpler thing we can do, which would reduce a lot of maintenance overhead for us -- Rather, than managing an entire sqlite db, we can just store the list of all tweeted out repositories in a text file, and store that somewhere (using an endpoint).

The code would be much more simpler, and robust. Let me know if you've any thoughts on this.

**Comment by sam95** [comment]
Hi @RJ722 , 

Yeah, we can do that. Storing it and reading it from a file would still work and would be much simpler than performing DB operations. But my thought is wouldn't we need a DB say even in the future? Like how @jaipradeesh suggested, we would want to tweet the DB occasionally after say every 'x' days. Also Sanket had mentioned a search feature on the dashboard for which we can re-use the DB and many more features to come.
And also managing an SQLite wouldn't be that hard. We need not setup a DB instance just for this. Managing the SQLite on the instance server where we've deployed the repo should be more than enough since there won't be heavy reads and writes. 
You can discuss it with @jaipradeesh and let me know either ways, Code changes to read and write to a json file should be much simpler and shouldn't take more than half a day for implementation from my side.

**Comment by RJ722** [comment]
Hey @sam95,

Thanks for your inputs, but I am still in favour of using a simple text based json file because of the reduced maintainence cost.

> Also Sanket had mentioned a search feature on the dashboard for which we can re-use the DB and many more features to come.

Since we already have structured information, I think we can use [aloglia] to implement search.

Reg. tweeting every 'x' days, when we store the file, we should also store some metadata, like the timestamp it was tweeted out.

**Comment by sam95** [comment]
Hi @RJ722,

Sure will have the updated PR soon.

**Comment by RJ722** [comment]
Hey @sam95, thanks for your time and work. I took inspiration from your code and addressed these changes in #52. Would you like to review those?

Closing this PR as inactive.

**Pull request closed without merging**